notes/core-abstractions.md

```diff
@@ -0,0 +1,97 @@
+# Core Abstractions
+
+This document describes the foundational abstractions in Shrink Ray.
+
+## ReductionProblem[T]
+
+The central abstraction is `ReductionProblem[T]` from `problem.py`. This represents a test-case reduction task over values of type T.
+
+### Key Properties
+
+- **current_test_case**: The best-known interesting test case
+- **is_interesting(test_case) -> bool**: Async predicate that tests if a candidate preserves the bug
+- **sort_key(test_case)**: Returns a comparable key for ordering - defaults to shortlex ordering
+- **size(test_case) -> int**: Returns the size of a test case (for metrics)
```

Comment:
Also used as a heuristic for how good a reduction pass is (size-reducing passes are much better than non-size reducing ones).